keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,492 @@
+Similar to the Ingress and Gateway APIs, this API will be dependent on
+implementations by controllers that are not included by default in Kubernetes.
+This could lead to confusion for users. We'll need to rely heavily on
+documentation for this feature, tracking all uses of official Kubernetes APIs
+that support ReferenceGrant in a central place.
+
+#### Potential for Variations Among Implementations
+Because this relies on each individual controller to implement the logic,
+it is possible that implementations may become inconsistent. To avoid that,
+we'll provide a standard library for implementing ReferenceGrant. We'll
+also strongly recommend that every API that relies on ReferenceGrant 
+includes robust conformance tests covering this functionality. Existing
+Gateway API conformance tests can serve as a model for this.
+
+#### Cross-Namespace References may Weaken Namespace Boundaries
+Although we believe that the handshake required for cross-namespace references
+with ReferenceGrant ensures these references will be safe, it does potentially
+weaken existing namespace boundaries. We believe ReferenceGrant will have a net
+benefit on the ecosystem as it will allow workloads, secrets, and configuration
+to be deployed in separate namespaces that more clearly match up with desired
+authorization.
+
+## Design Details
+
+### General Notes
+
+#### `ReferenceGrant` is half of a handshake
+
+When thinking about ReferenceGrant, it is important to remember that it does not
+do anything by itself. It *Grants* the *possibility* of making a *Reference*
+across namespaces. It's intended that _another object_ (that is, the From object)
+complete the handshake by creating a reference to the referent object (the To
+object).
+
+#### `Kind` vs `Resource`
+
+When creating a metaresource (that is, a resource that targets other resources)
+like ReferenceGrant, it's important to consider if the metaresource uses the more
+common `Kind` or the more correct `Resource`.
+
+When designing the Gateway API in general, we considered this quite a bit (and
+@robscott even ended up making https://github.com/kubernetes/community/pull/5973
+to clarify the API conventions), but in the end decided to use Kind, to improve
+the user experience. That is, it's easier users to take the value from the `kind`
+field at the top of the YAML they are already using, and put it straight into
+these fields, rather than needing to do a kind-resource lookup for every user's
+interaction with the API.
+
+Secondly, as in the Ingress API, the target types must be clearly defined by
+each implementation, so there should be no opportunity for ambiguity.
+
+Lastly, this is only half of a handshake, so the From referent must also both
+support referencing the To referent, and must reference the one in the To field.
+
+With all of that in mind, we felt that the benefits in terms of specificity
+outweighed the costs in terms of user pain to use the more correct `Resource`.
```

Comment:
Using `kind` instead of `resource` means that any authorization verification that the user creating/modifying the ReferenceGrant actually has permission to access the object they are granting access to has to do a REST mapping lookup, which could fail. Examples at:
* https://github.com/kubernetes/kubernetes/blob/release-1.26/pkg/registry/admissionregistration/validatingadmissionpolicy/authz.go#L72-L83
* https://github.com/kubernetes/kubernetes/blob/release-1.26/plugin/pkg/admission/gc/gc_admission.go#L206-L215

That should at least be mentioned as a downside of using `kind`, and discuss how those error cases should be handled

Reply:
Yeah, I think `Kind` is insufficient here.  It needs to be `Resource`.

Within a very limited domain, Kind is OK - the implementing controller knows that Group-Kind -> Resource is bounded and enumerable (and usually in a group under the same control).  E.g. a Foo Gateway impl allows Kind "FooRoute".  

In the case of RefGrant, a Kind would seem to mean _any_ resource which uses that Kind (I hate that we have this distinction, TBH, but...), and I doubt that is the sort of ambiguity we want in an authz API.

Reply:
Okay, I guess I've lost this one. 😄 

The one advantage I can see of using Resource rather than Kind is that resources are plural, so it will end up more clear that you're granting access from more than one thing.

So, instead of allowing access to Gateways using this one (as we do now in Gateway API):
```yaml
    from:
      # Note that in Gateway API, Group is currently defaulted
      # to this, which means you to explicitly set the group to
      # the empty string for Core resources. We should definitely
      # change this.
      - group: "gateway.networking.kubernetes.io"
        kind: Gateway
        namespace: foo
```

It will be this one:
```yaml
    from:
      # Assuming that we leave the default for Group to the empty
      # string, so that Core objects don't need additional config.
      - group: "gateway.networking.kubernetes.io" 
        resource: gateways
        namespace: foo
```

Which implies that it's _all_ Gateways in the namespace better.

It does suck a bit that the docstring for the field will need to remind people about kind/resource distinctions, but I agree that we need the extra caution here. Sigh.

Reply:
I'll do this update tomorrow my time (finishing up for the day soon).